File: lcof-python/剑指offer/ListNode/6_链表第一个公共节点.py
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

#
# 
# @param pHead1 ListNode类 
# @param pHead2 ListNode类 
# @return ListNode类
#
# 哈希实现
class Solution:
    def FindFirstCommonNode(self , pHead1 , pHead2 ):
        # write code here
        # 需要额外空间
        if pHead1 is None or pHead2 is None:
            return None
        hash_table = set()
        head1 = pHead1
        head2 = pHead2
        while pHead1:
            hash_table.add(pHead1)
            pHead1 = pHead1.next
        
        while pHead2:
            if pHead2 in hash_table:
                return pHead2
            pHead2 = pHead2.next
        return None

File: lcof-cpp/offer/1_ListNode/6_链表第一个公共节点.cpp
/*
struct ListNode {
    int val;
    struct ListNode *next;
    ListNode(int x) :
            val(x), next(NULL) {
    }
};*/
class Solution
{
public:
    ListNode *FindFirstCommonNode(ListNode *pHead1, ListNode *pHead2)
    {
        if (!pHead1)
            return nullptr;
        if (!pHead2)
            return nullptr;
        ListNode *M = pHead1;
        ListNode *N = pHead2;
        while (M != N)
        {
            M = M ? M->next : pHead2; // 如果M为空才重置，否则正常移动
            N = N ? N->next : pHead1; // 如果N为空才重置，否则正常移动
        }
        return M;
    }
};
